# Voting Classifier Lab

### Intro and objectives


### In this lab you will learn:
1. a basic example of voting-based classifier

### What I hope you'll get out of this lab
* Worked Examples
* How to interpret the results obtained

In [ ]:
import sys

assert sys.version_info >= (3, 7)

In [1]:
from packaging import version
import sklearn

assert version.parse(sklearn.__version__) >= version.parse("1.0.1")

### In this lab we will learn how build an ensemble classifier based on the voting technique

In [3]:
from sklearn.datasets import make_moons
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC

## Let's create some synthetic data

In [4]:
X, y = make_moons(n_samples=500, noise=0.30, random_state=42)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

## Let's create and train a voting classifier composed of three diverse classifiers (logistic regression, random forest, SVC)

In [5]:
voting_clf = VotingClassifier(
    estimators=[
        ('lr', LogisticRegression(random_state=42)),
        ('rf', RandomForestClassifier(random_state=42)),
        ('svc', SVC(random_state=42))
    ]
)
voting_clf.fit(X_train, y_train)

VotingClassifier(estimators=[('lr', LogisticRegression(random_state=42)),
                             ('rf', RandomForestClassifier(random_state=42)),
                             ('svc', SVC(random_state=42))])

## Let's run some predictions

In [23]:
X_test[0:10]

array([[ 0.50169252,  0.21717211],
       [ 0.22740469,  0.5899372 ],
       [ 0.26816663,  0.43111374],
       [ 1.31270833, -0.59523574],
       [ 0.41835564, -0.3826774 ],
       [ 0.53680639, -0.32084028],
       [-0.19975183,  1.43509879],
       [-0.85385457,  0.71924064],
       [-0.57050579,  0.00411171],
       [-0.67806633,  0.74946744]])

In [22]:
y_test[0:10]

array([1, 1, 1, 1, 1, 1, 0, 0, 0, 0])

In [31]:
voting_clf.predict([X_test[0]])

array([1])

In [32]:
voting_clf.predict([X_test[9]])

array([0])

In [34]:
voting_clf.predict(X_test[0:10])

array([1, 0, 0, 1, 1, 1, 0, 0, 0, 0])

## Let's evauate the performance

In [35]:
for name, clf in voting_clf.named_estimators_.items():
    print(name, "=", clf.score(X_test, y_test))

lr = 0.864
rf = 0.896
svc = 0.896


In [36]:
voting_clf.score(X_test, y_test)

0.912

#### From the previous results we observe that the voting classifier performs better (0.912) than each individual classifier (0.864, 0.896, 0.896).